doc/developer/design/20220203_connectors.md

```diff
@@ -0,0 +1,160 @@
+configuration parameters. The current design of `CREATE SOURCE` and `CREATE
+SINK` make this a verbose affair. The problem is particularly acute with
+Avro-formatted Kafka sources that configure authentication:
+
+```sql
+CREATE SOURCE kafka1
+FROM KAFKA BROKER 'kafka:9092' TOPIC 'top1' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password',
+)
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+
+CREATE SOURCE kafka2
+FROM KAFKA BROKER 'kafka:9092' TOPIC 'top2' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password'
+)
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+```
+
+These two source definition differ only in their topic specification, but
+must duplicate eight other configuration parameters.
+
+With the connectors proposed in this document, the `CREATE SOURCE` can instead
+share all the relevant configuration:
+
+```sql
+CREATE CONNECTOR kafka FOR
+KAFKA BROKER 'kafka:9092' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password'
+);
+
+CREATE CONNECTOR schema_registry FOR
+CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+
+CREATE SOURCE kafka1
+FROM KAFKA CONNECTOR kafka TOPIC 'top1'
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTOR schema_registry;
+
+CREATE SOURCE kafka2
+FROM KAFKA CONNECTOR kafka TOPIC 'top2'
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTOR schema_registry;
+```
+
+## Design
+
+### SQL syntax and semantics
+
+The eventual goal is for connector create syntax to come in type specific variants and not include `WITH` options except in cases of client implementation specific details; however, the initial implementation will preserve existing syntax for connectors to avoid entangling separating connectors into their own objects from reworking the overall connection option syntax.
+
+All Sources in the catalog will have a reference to their connector which will also indicate if the connector is implicit or explicit. Implicit connectors are created when a traditional `CREATE SOURCE` command is executed and are invisible unless the catalog is directly inspected, i.e. `SHOW CREATE SOURCE` for a source with an implicit connector will return the same output as if connectors did not exist. Explicit connectors are created with `CREATE CONNECTOR` commands and used by specifying the connector in the `CREATE SOURCE` statement. 
+
+## Implementation Phases 
+
+1. Support implicit and explicit connectors using otherwise identical syntax to now (`WITH` options, etc)
```

Comment:
I think I might have mentioned this, but I propose we make `CREATE CONNECTOR` experimental to start, so that we can defer the decision on whether to ship it with or without the syntax redesign.

Reply:
Awesome, I will add the callout that the experimental flag will cover both implicit and explicit behavior. Is it reasonable to make the read support for the experimental feature non-experimental? i.e. you have to opt-in to write the new type, but we'll read and do the right thing with either format at all times?

Reply:
> Awesome, I will add the callout that the experimental flag will cover both implicit and explicit behavior. Is it reasonable to make the read support for the experimental feature non-experimental? i.e. you have to opt-in to write the new type, but we'll read and do the right thing with either format at all times?

Yes, 100%! That is the usual approach we take and it works well.